hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java

```diff
@@ -926,18 +950,24 @@ protected Record putNewRecord(Data key, Object oldValue, Object newValue, long t
                                   EntryEventType entryEventType, boolean store,
                                   boolean backup) {
         Record record = createRecord(key, newValue, now);
-        if (mapDataStore != EMPTY_MAP_DATA_STORE && store) {
-            putIntoMapStore(record, key, newValue, ttl, maxIdle, now, transactionId);
-        }
-        storage.put(key, record);
-        expirySystem.add(key, ttl, maxIdle, expiryTime, now, now);
+        // Ensure the record is kept in-memory until being released
+        storage.pinRecord(record);
+        try {
+            if (mapDataStore != EMPTY_MAP_DATA_STORE && store) {
+                putIntoMapStore(record, key, newValue, ttl, maxIdle, now, transactionId);
+            }
+            storage.put(key, record);
+            expirySystem.add(key, ttl, maxIdle, expiryTime, now, now);
 
-        if (entryEventType == EntryEventType.LOADED) {
-            mutationObserver.onLoadRecord(key, record, backup);
-        } else {
-            mutationObserver.onPutRecord(key, record, oldValue, backup);
+            if (entryEventType == EntryEventType.LOADED) {
+                mutationObserver.onLoadRecord(key, record, backup);
+            } else {
+                mutationObserver.onPutRecord(key, record, oldValue, backup);
+            }
+            return record;
+        } finally {
+            storage.unpinRecord(record);
```

Comment:
dummy question - the returned record is not pinned, which might be accessed by some other methods. Is it safe?

Reply:
The pinning mechanism guarantees that the record is kept in memory region (not just in a detached memory chunk). Since unpin doesn't touch the epoch, it is still protected by the current epoch.

Reply:
i see, thanks :)